inference/scripts/test-wasm.py

```diff
@@ -0,0 +1,91 @@
+INFERENCE_PATH = os.path.dirname(SCRIPTS_PATH)
+BUILD_PATH = os.path.join(INFERENCE_PATH, "build-wasm")
+WASM_PATH = os.path.join(INFERENCE_PATH, "wasm")
+WASM_TESTS_PATH = os.path.join(WASM_PATH, "tests")
+GENERATED_PATH = os.path.join(WASM_TESTS_PATH, "generated")
+MODELS_PATH = os.path.join(WASM_TESTS_PATH, "models")
+WASM_ARTIFACT = os.path.join(BUILD_PATH, "bergamot-translator.wasm")
+JS_ARTIFACT = os.path.join(BUILD_PATH, "bergamot-translator.js")
+JS_ARTIFACT_HASH = os.path.join(GENERATED_PATH, "bergamot-translator.js.sha256")
+
+
+def calculate_sha256(file_path):
+    sha256_hash = hashlib.sha256()
+    with open(file_path, "rb") as f:
+        for byte_block in iter(lambda: f.read(4096), b""):
+            sha256_hash.update(byte_block)
+    return sha256_hash.hexdigest()
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Test WASM by building and handling artifacts.",
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+
+    parser.add_argument("--clobber", action="store_true", help="Clobber the build artifacts")
+    parser.add_argument(
+        "--debug",
+        action="store_true",
+        help="Build with debug symbols, useful for profiling",
+    )
+    parser.add_argument(
+        "-j",
+        type=int,
+        help="Number of cores to use for building (default: all available cores)",
+    )
+    args = parser.parse_args()
+
+    build_wasm_script = os.path.join(SCRIPTS_PATH, "build-wasm.py")
+    build_command = [sys.executable, build_wasm_script]
+    if args.clobber:
+        build_command.append("--clobber")
+    if args.debug:
+        build_command.append("--debug")
+    if args.j:
+        build_command.extend(["-j", str(args.j)])
+
+    print("\n🚀 Starting build-wasm.py")
+    subprocess.run(build_command, check=True)
+
+    print("\n📁 Copying generated build artifacts to the WASM test directory\n")
+
+    os.makedirs(GENERATED_PATH, exist_ok=True)
+    shutil.copy2(WASM_ARTIFACT, GENERATED_PATH)
+    shutil.copy2(JS_ARTIFACT, GENERATED_PATH)
+
+    print(f"   Copied the following artifacts to {GENERATED_PATH}:")
+    print(f"     - {JS_ARTIFACT}")
+    print(f"     - {WASM_ARTIFACT}")
+
+    print(f"\n🔑 Calculating SHA-256 hash of {JS_ARTIFACT}\n")
+    hash_value = calculate_sha256(JS_ARTIFACT)
+    with open(JS_ARTIFACT_HASH, "w") as hash_file:
+        hash_file.write(f"{hash_value}  {os.path.basename(JS_ARTIFACT)}\n")
+    print(f"   Hash of {JS_ARTIFACT} written to")
+    print(f"   {JS_ARTIFACT_HASH}")
+
+    print("\n📂 Decompressing model files required for WASM testing\n")
+    subprocess.run(["gzip", "-dkrf", MODELS_PATH], check=True)
+    print(f"   Decompressed models in {MODELS_PATH}\n")
+
+    print("\n🔧 Installing npm dependencies for WASM JS tests\n")
+    subprocess.run(["npm", "install"], cwd=WASM_TESTS_PATH, check=True)
+
+    print("\n📊 Running Translations WASM JS tests\n")
+    subprocess.run(["npm", "run", "test"], cwd=WASM_TESTS_PATH, check=True)
```

Comment:
Suggestion: In CI I would rather have all this output actually show up so it's easy to deal with test failures. In jest you can have each test that is run enumerated in the log output. It's not clear from the task output that the tests actually ran, or what tests were run.

https://github.com/mozilla/translations/pull/927/checks?check_run_id=32892931631

```
[task 2024-11-12T23:03:00.175Z] 📂 Decompressing model files required for WASM testing
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z]    Decompressed models in /builds/worker/checkouts/vcs/inference/wasm/tests/models
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 🔧 Installing npm dependencies for WASM JS tests
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 📊 Running Translations WASM JS tests
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] ✅ test-wasm.py completed successfully.
[task 2024-11-12T23:03:00.175Z] 
[taskcluster 2024-11-12 23:03:00.768Z] === Task Finished ===
```

Reply:
Hmm, you know this is odd. 

Locally, the output looks like this:

```
🚀 Starting build-wasm.py

🏗️  Build directory /home/nordzilla/src/translations/inference/build-wasm already exists.

   Pass the --clobber flag to rebuild if desired.

📁 Copying generated build artifacts to the WASM test directory

   Copied the following artifacts to /home/nordzilla/src/translations/inference/wasm/tests/generated:
     - /home/nordzilla/src/translations/inference/build-wasm/bergamot-translator.js
     - /home/nordzilla/src/translations/inference/build-wasm/bergamot-translator.wasm

🔑 Calculating SHA-256 hash of /home/nordzilla/src/translations/inference/build-wasm/bergamot-translator.js

   Hash of /home/nordzilla/src/translations/inference/build-wasm/bergamot-translator.js written to
   /home/nordzilla/src/translations/inference/wasm/tests/generated/bergamot-translator.js.sha256

📂 Decompressing model files required for WASM testing

   Decompressed models in /home/nordzilla/src/translations/inference/wasm/tests/models


🔧 Installing npm dependencies for WASM JS tests

(node:111180) ExperimentalWarning: CommonJS module /home/nordzilla/.nvm/versions/node/v23.1.0/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /home/nordzilla/.nvm/versions/node/v23.1.0/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)

up to date, audited 135 packages in 1s

38 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

📊 Running Translations WASM JS tests


> wasm-tests@1.0.0 test
> vitest --run


 RUN  v2.1.4 /home/nordzilla/src/translations/inference/wasm/tests

 ✓ test-cases/translate-html-no-pivot.test.mjs (4) 1578ms
   ✓ HTML Non-Pivot Translations (4) 1578ms
     ✓ (es -> en): Translate "<b>El perro</b> azul." 374ms
     ✓ (en -> es): Translate "<b>The blue</b> dog." 423ms
     ✓ (fr -> en): Translate "<b>Le chien</b> bleu." 406ms
     ✓ (en -> fr): Translate "<b>The blue</b> dog." 374ms
 ✓ test-cases/translate-html-with-pivot.test.mjs (2) 1271ms
   ✓ HTML Pivot Translations (2) 1271ms
     ✓ (es -> fr): Translate "<b>El perro</b> azul." 610ms
     ✓ (fr -> es): Translate "<b>Le chien</b> bleu." 660ms
 ✓ test-cases/translate-plain-text-no-pivot.test.mjs (4) 1601ms
   ✓ Plain-Text Non-Pivot Translations (4) 1601ms
     ✓ (es -> en): Translate "Hola mundo" 384ms
     ✓ (en -> es): Translate "Hello world" 415ms
     ✓ (fr -> en): Translate "Bonjour le monde" 414ms
     ✓ (en -> fr): Translate "Hello world" 387ms
 ✓ test-cases/translate-plain-text-with-pivot.test.mjs (2) 1263ms
   ✓ Plain-Text Pivot Translations (2) 1263ms
     ✓ (es -> fr): Translate "El perro azul." 592ms
     ✓ (fr -> es): Translate "Le chien bleu." 671ms

 Test Files  4 passed (4)
      Tests  12 passed (12)
   Start at  12:32:46
   Duration  1.94s (transform 42ms, setup 0ms, collect 98ms, tests 5.71s, environment 1ms, prepare 230ms)


✅ test-wasm.py completed successfully.
```

But if you look at the link you pasted for the logs, the output from anything related to `node` is inserted above the entire Python script's output:

```
[task 2024-11-12T23:02:55.451Z] (node:8372) ExperimentalWarning: CommonJS module /builds/worker/.nvm/versions/node/v23.1.0/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /builds/worker/.nvm/versions/node/v23.1.0/lib/node_modules/npm/node_modules/supports-color/index.js using require().
[task 2024-11-12T23:02:55.451Z] Support for loading ES Module in require() is an experimental feature and might change at any time
[task 2024-11-12T23:02:55.451Z] (Use `node --trace-warnings ...` to show where the warning was created)
[task 2024-11-12T23:02:57.257Z] 
[task 2024-11-12T23:02:57.257Z] added 134 packages, and audited 135 packages in 2s
[task 2024-11-12T23:02:57.258Z] 
[task 2024-11-12T23:02:57.258Z] 38 packages are looking for funding
[task 2024-11-12T23:02:57.258Z]   run `npm fund` for details
[task 2024-11-12T23:02:57.259Z] 
[task 2024-11-12T23:02:57.259Z] found 0 vulnerabilities
[task 2024-11-12T23:02:57.385Z] 
[task 2024-11-12T23:02:57.385Z] > wasm-tests@1.0.0 test
[task 2024-11-12T23:02:57.385Z] > vitest --run
[task 2024-11-12T23:02:57.385Z] 
[task 2024-11-12T23:02:57.697Z] 
[task 2024-11-12T23:02:57.697Z]  RUN  v2.1.4 /builds/worker/checkouts/vcs/inference/wasm/tests
[task 2024-11-12T23:02:57.697Z] 
[task 2024-11-12T23:02:59.706Z]  ✓ test-cases/translate-plain-text-with-pivot.test.mjs  (2 tests) 1715ms
[task 2024-11-12T23:02:59.706Z]    ✓ Plain-Text Pivot Translations > (es -> fr): Translate "El perro azul." 853ms
[task 2024-11-12T23:02:59.706Z]    ✓ Plain-Text Pivot Translations > (fr -> es): Translate "Le chien bleu." 860ms
[task 2024-11-12T23:02:59.758Z]  ✓ test-cases/translate-html-with-pivot.test.mjs  (2 tests) 1768ms
[task 2024-11-12T23:02:59.759Z]    ✓ HTML Pivot Translations > (es -> fr): Translate "<b>El perro</b> azul." 882ms
[task 2024-11-12T23:02:59.759Z]    ✓ HTML Pivot Translations > (fr -> es): Translate "<b>Le chien</b> bleu." 885ms
[task 2024-11-12T23:03:00.093Z]  ✓ test-cases/translate-plain-text-no-pivot.test.mjs  (4 tests) 2103ms
[task 2024-11-12T23:03:00.093Z]    ✓ Plain-Text Non-Pivot Translations > (es -> en): Translate "Hola mundo" 545ms
[task 2024-11-12T23:03:00.093Z]    ✓ Plain-Text Non-Pivot Translations > (en -> es): Translate "Hello world" 515ms
[task 2024-11-12T23:03:00.093Z]    ✓ Plain-Text Non-Pivot Translations > (fr -> en): Translate "Bonjour le monde" 544ms
[task 2024-11-12T23:03:00.093Z]    ✓ Plain-Text Non-Pivot Translations > (en -> fr): Translate "Hello world" 498ms
[task 2024-11-12T23:03:00.129Z]  ✓ test-cases/translate-html-no-pivot.test.mjs  (4 tests) 2136ms
[task 2024-11-12T23:03:00.129Z]    ✓ HTML Non-Pivot Translations > (es -> en): Translate "<b>El perro</b> azul." 549ms
[task 2024-11-12T23:03:00.129Z]    ✓ HTML Non-Pivot Translations > (en -> es): Translate "<b>The blue</b> dog." 555ms
[task 2024-11-12T23:03:00.129Z]    ✓ HTML Non-Pivot Translations > (fr -> en): Translate "<b>Le chien</b> bleu." 532ms
[task 2024-11-12T23:03:00.129Z]    ✓ HTML Non-Pivot Translations > (en -> fr): Translate "<b>The blue</b> dog." 498ms
[task 2024-11-12T23:03:00.152Z] 
[task 2024-11-12T23:03:00.153Z]  Test Files  4 passed (4)
[task 2024-11-12T23:03:00.153Z]       Tests  12 passed (12)
[task 2024-11-12T23:03:00.154Z]    Start at  23:02:57
[task 2024-11-12T23:03:00.154Z]    Duration  2.45s (transform 46ms, setup 0ms, collect 98ms, tests 7.72s, environment 1ms, prepare 325ms)
[task 2024-11-12T23:03:00.154Z] 
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 🚀 Starting build-wasm.py
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 📁 Copying generated build artifacts to the WASM test directory
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z]    Copied the following artifacts to /builds/worker/checkouts/vcs/inference/wasm/tests/generated:
[task 2024-11-12T23:03:00.175Z]      - /builds/worker/checkouts/vcs/inference/build-wasm/bergamot-translator.js
[task 2024-11-12T23:03:00.175Z]      - /builds/worker/checkouts/vcs/inference/build-wasm/bergamot-translator.wasm
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 🔑 Calculating SHA-256 hash of /builds/worker/checkouts/vcs/inference/build-wasm/bergamot-translator.js
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z]    Hash of /builds/worker/checkouts/vcs/inference/build-wasm/bergamot-translator.js written to
[task 2024-11-12T23:03:00.175Z]    /builds/worker/checkouts/vcs/inference/wasm/tests/generated/bergamot-translator.js.sha256
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 📂 Decompressing model files required for WASM testing
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z]    Decompressed models in /builds/worker/checkouts/vcs/inference/wasm/tests/models
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 🔧 Installing npm dependencies for WASM JS tests
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 📊 Running Translations WASM JS tests
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] 
[task 2024-11-12T23:03:00.175Z] ✅ test-wasm.py completed successfully.
```

I don't know why it's out of order in the taskcluster logs. Maybe just a Taskcluster oddity (cc @bhearsum)? 

In any case, once we have more tests, I think it will be harder to miss. It is there, though.